src/ifcwrap-csharp/IfcOpenShell.Net/IfcOpenShell.Net/EntityInstanceExtensions.cs

```diff
@@ -0,0 +1,391 @@
+            {
+                foreach (var def in definers)
+                {
+                    if (def.Is("IfcRelDefinesByType") && def.TryGetAttributeAsEntity("RelatingType", out var type))
+                    {
+                        return type;
+                    }
+                }
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Return the unique id of the element in the ifc file.
+        /// However, this id only exists for ifc objects that inherit from IfcRoot, so may be null.
+        /// </summary>
+        /// <param name="element"></param>
+        /// <returns></returns>
+        public static string? Guid(this EntityInstance element)
+        {
+            if (element.Is("IfcRoot"))
+            {
+                return element.GetAttributeAsString("GlobalId");
+            }
+            return null;
+        } 
+        
+        
+        /// <summary>
+        /// Return a dictionary of the entity_instance's properties (Python and IFC) and their values.
+        /// </summary>
+        /// <param name="instance"></param>
+        /// <param name="includeIdentifier">Whether or not to include the STEP numerical identifier</param>
+        /// <param name="recursive">Whether or not to convert referenced IFC elements into dictionaries too. All attributes also apply recursively</param>
+        /// <param name="scalarOnly">Filters out all values that are IFC instances</param>
+        /// <param name="ignore">A list of attribute names to ignore</param>
+        /// <returns>A dictionary of properties and their corresponding values</returns>
+        public static IReadOnlyDictionary<string, ArgumentResult> GetInfo(this EntityInstance instance, bool includeIdentifier=true, bool recursive = false, bool scalarOnly=false, IEnumerable<string>? ignore=null)
+        {
+            var ignoreHashSet = ignore?.ToHashSet() ?? new HashSet<string>();
+            
+            var result = new Dictionary<string, ArgumentResult>();
+            try
+            {
+                if(includeIdentifier) result.Add("id", new ArgumentResult.FromInt(instance.id()));
+                result.Add("type", new ArgumentResult.FromString(instance.Is()));
+            }
+            catch (Exception e)
+            {
+                // TODO: logging instead
+                throw;
+            }
+
+            var attributeNames = instance.get_attribute_names();
+            
+            foreach (var attributeName in attributeNames)
+            {
+                try
+                {
+                    if (ignoreHashSet.Contains(attributeName))
+                        continue;
+                    if(!instance.TryGetAttributeAsEntity(attributeName, out var attribute)) 
+                        continue;
+                    var shouldInclude = true;
+                    if (recursive || scalarOnly)
+                        throw new NotImplementedException();
```

Comment:
Add a exception body so that when we catch it, we know that this is one of *our* NIEs which we purpusfully omitted